apps/progs.c

```diff
@@ -13,388 +13,388 @@
-    {FT_general, "rehash", rehash_main, rehash_options},
-    {FT_general, "req", req_main, req_options},
-    {FT_general, "rsa", rsa_main, rsa_options},
+    {FT_general, "pkcs12", pkcs12_main, pkcs12_options, NULL},
+#endif
+    {FT_general, "pkcs7", pkcs7_main, pkcs7_options, NULL},
+    {FT_general, "pkcs8", pkcs8_main, pkcs8_options, NULL},
+    {FT_general, "pkey", pkey_main, pkey_options, NULL},
+    {FT_general, "pkeyparam", pkeyparam_main, pkeyparam_options, NULL},
+    {FT_general, "pkeyutl", pkeyutl_main, pkeyutl_options, NULL},
+    {FT_general, "prime", prime_main, prime_options, NULL},
+    {FT_general, "provider", provider_main, provider_options, NULL},
+    {FT_general, "rand", rand_main, rand_options, NULL},
+    {FT_general, "rehash", rehash_main, rehash_options, NULL},
+    {FT_general, "req", req_main, req_options, NULL},
+    {FT_general, "rsa", rsa_main, rsa_options, NULL},
 #ifndef OPENSSL_NO_RSA
-    {FT_general, "rsautl", rsautl_main, rsautl_options},
+    {FT_general, "rsautl", rsautl_main, rsautl_options, NULL},
 #endif
 #ifndef OPENSSL_NO_SOCK
-    {FT_general, "s_client", s_client_main, s_client_options},
+    {FT_general, "s_client", s_client_main, s_client_options, NULL},
 #endif
 #ifndef OPENSSL_NO_SOCK
-    {FT_general, "s_server", s_server_main, s_server_options},
+    {FT_general, "s_server", s_server_main, s_server_options, NULL},
 #endif
 #ifndef OPENSSL_NO_SOCK
-    {FT_general, "s_time", s_time_main, s_time_options},
+    {FT_general, "s_time", s_time_main, s_time_options, NULL},
 #endif
-    {FT_general, "sess_id", sess_id_main, sess_id_options},
-    {FT_general, "smime", smime_main, smime_options},
-    {FT_general, "speed", speed_main, speed_options},
-    {FT_general, "spkac", spkac_main, spkac_options},
+    {FT_general, "sess_id", sess_id_main, sess_id_options, NULL},
+    {FT_general, "smime", smime_main, smime_options, NULL},
+    {FT_general, "speed", speed_main, speed_options, NULL},
+    {FT_general, "spkac", spkac_main, spkac_options, NULL},
 #ifndef OPENSSL_NO_SRP
-    {FT_general, "srp", srp_main, srp_options},
+    {FT_general, "srp", srp_main, srp_options, NULL},
 #endif
-    {FT_general, "storeutl", storeutl_main, storeutl_options},
+    {FT_general, "storeutl", storeutl_main, storeutl_options, NULL},
 #ifndef OPENSSL_NO_TS
-    {FT_general, "ts", ts_main, ts_options},
+    {FT_general, "ts", ts_main, ts_options, NULL},
 #endif
-    {FT_general, "verify", verify_main, verify_options},
-    {FT_general, "version", version_main, version_options},
-    {FT_general, "x509", x509_main, x509_options},
-#ifndef OPENSSL_NO_MD2
-    {FT_md, "md2", dgst_main},
+    {FT_general, "verify", verify_main, verify_options, NULL},
+    {FT_general, "version", version_main, version_options, NULL},
+    {FT_general, "x509", x509_main, x509_options, NULL},
+#if !defined(OPENSSL_NO_MD2) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+    {FT_md, "md2", dgst_main, NULL, DEPRECATED_NO_ALTERNATIVE},
```

Comment:
Why are we deprecating these digests? Our strategy so far has been to deprecate APIs. In this PR we introduce the concept of also deprecating apps that *use* those APIs. But this seems to be going much further by deprecating *algorithms*. So far we haven't deprecated any algorithms - only moved them to the legacy provider.

Reply:
Hmmm, the algorithms as command get the EVP_MD or EVP_CIPHER via EVP_MD_by_digestname() and EVP_CIPHER_by_ciphername(), no? Doesn't deprecation come into play there?

Reply:
Does it? We're not deprecating those functions are we?

Reply:
I admit that my recollection is vague, gimme a moment or two to check things out, yeah?